clippy_lints/src/matches/redundant_pattern_match.rs

```diff
@@ -199,36 +201,61 @@ fn find_sugg_for_if_let<'tcx>(
 }
 
 pub(super) fn check_match<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>, op: &Expr<'_>, arms: &[Arm<'_>]) {
-    if let [arm1, arm2] = arms
-        && arm1.guard.is_none()
-        && arm2.guard.is_none()
-    {
-        let node_pair = (&arm1.pat.kind, &arm2.pat.kind);
+    if arms.len() == 2 {
+        let node_pair = (&arms[0].pat.kind, &arms[1].pat.kind);
 
-        if let Some(good_method) = found_good_method(cx, arms, node_pair) {
+        if let Some((good_method, maybe_guard)) = found_good_method(cx, arms, node_pair) {
             let span = is_expn_of(expr.span, "matches").unwrap_or(expr.span.to(op.span));
             let result_expr = match &op.kind {
                 ExprKind::AddrOf(_, _, borrowed) => borrowed,
                 _ => op,
             };
+            let mut sugg = format!("{}.{good_method}", snippet(cx, result_expr.span, "_"));
+
+            if let Some(guard) = maybe_guard {
+                let Guard::If(guard) = *guard else { return }; // `...is_none() && let ...` is a syntax error
+
+                // wow, the HIR for match guards in `PAT if let PAT = expr && expr => ...` is annoying!
+                // `guard` here is `Guard::If` with the let expression somewhere deep in the tree of exprs,
+                // counter to the intuition that it should be `Guard::IfLet`, so we need another check
+                // to see that there aren't any let chains anywhere in the guard, as that would break
+                // if we suggest `t.is_none() && (let X = y && z)` for:
+                // `match t { None if let X = y && z => true, _ => false }`
```

Comment:
I feel like this comment might be a bit confusing. If you ( the reviewer(s) ) already understand what this means, feel free to ignore.
Looking at the HIR here: https://godbolt.org/z/8413T1rjP (line 521-522), the guard is not `Guard::IfLet`, but `Guard::If`, with the `let` expression being potentially somewhere deep in the expression tree, so just checking `Guard::If` is not enough -- it may still be an if let guard despite the guard not being `Guard::IfLet`

Reply:
This is perhaps because it's desugared to something akin a let-chain, or at least very similar, which is like `And<Let, And<Expr, And<Expr, Let>>>` or something in the HIR, with `Expr` being `if` and `Let` being `let`, so you can probably check the first in `Binary::And` afaik but perhaps this would have issues on something like `if true && let Some(x) = x`, I'm not sure. Better safe than sorry though so the current way is fine